### PR TITLE
log source code instead of function pointers on error

### PR DESCRIPTION
**Pull request by McBen**
Function pointers doesn't give much hints where they came from.
Use 'toString' to see the source code of the bad function.

### CODE REVIEW

**Comment by Aradiv** [comment]
why don't you just use breakpoints and the debugger?

just printing the entire source also isn't very helpful as we already have the plugin name and can just look into the plugin source.
If you really want to have the code limit the output to only the line + surroundings of the line where the error occurred

**Comment by McBen** [comment]
The current message gives no hint which plugin is raising the error.
grafik
(source link goes to the exception handler 'runHook')

> why don't you just use breakpoints and the debugger?

Sometimes a error occures for unknown reason. Like a plugin failed when a special (unkown) portal is loaded in link-zoom.

> just printing the entire source also isn't very helpful as we already have the plugin nam

as said, I don't see any plugin name here.  It only logs the code of error function.

**Comment by McBen** [comment]
The current message gives no hint which plugin is raising the error. 
 
FF: (with PR)
grafik

/edit..ah, I see chrome does a better job here. Firefox lacks the call stack of the exception.
grafik


> why don't you just use breakpoints and the debugger? 
 
Sometimes an error occures for unknown reason. Like a plugin failed when a special (unkown) portal is loaded in link-zoom. 
 
> just printing the entire source also isn't very helpful as we already have the plugin nam 
 
as said, I don't see any plugin name here.  It only logs the code of error function.

**Comment by Aradiv** [comment]
Okay so logging the plugin name can actually be helpfull but i still don't see how logging the entire source is helpful.

Especially those unknown reasons are the point where the debugger is very helpful as you can inspect the entire callstack when the error occurs. Just enable "pause on exceptions" and you can go to the entire callstack with all the values.

**Comment by McBen** [comment]
> logging the plugin name can actually be helpfull 

Logging the plugin and function name would be great. But I've no idea how to archieve this.
So this is my approach.

**Comment by johnd0e** [comment]
> But I've no idea how to archieve this.

We have to change plugins api in order to achieve that.

File: core/code/boot.js [hunk 1]
```diff
@@ -685,14 +685,14 @@ function prepPluginsToLoad() {
...
-    if (!setup) {
+  function safeSetup(setup) {
+    if (typeof setup !== 'function') {
       log.warn('plugin must provide setup function');
       return;
     }
     var info = setup.info;
     if (typeof info !== 'object') {
-      log.warn('plugin does not have proper wrapper:',setup);
+      log.warn('plugin does not have proper wrapper:', { function: setup, info: setup.info, source: setup.toString() });
```
Comment (johnd0e) [inline core/code/boot.js:695]: BTW, this is not needed in Chrome, as it already shows function test, and allows to jump to `setup` definition.

File: core/code/hooks.js [hunk 2]
```diff
@@ -74,7 +74,7 @@ window.runHooks = function(event, data) {
         return false;  //break from $.each
       }
     } catch (e) {
-      log.error('error running hook '+event+', error: '+e);
+      log.error('error running hook', { event: event, error: e, source: callback.toString(), data: data });
```
Comment (johnd0e) [inline core/code/hooks.js]: Have you tried that in IITCm?
I suspect that it is unable to show objects in console.

File: core/code/boot.js (hunk 1 above)
Comment (McBen) [reply core/code/boot.js:695]: My main system is Firefox so I didn't recognized that before.

see https://github.com/IITC-CE/ingress-intel-total-conversion/pull/385#issuecomment-676552427

File: core/code/hooks.js (hunk 2 above)
Comment (McBen) [reply core/code/hooks.js]: It will work in IITCm. It won't work in IITCm console window.

**Pull request closed without merging**